File: android/app/src/main/java/com/skyai/app/data/model/Flight.kt
package com.skyai.app.data.model

import com.google.gson.annotations.SerializedName

// ===== Enums =====
//
// All wire values mirror backend Pydantic enums (see backend/models.py). When
// the backend ships UPPERCASE strings, so do we — Gson's @SerializedName is
// case-sensitive and bare property names default to camelCase, so omitting
// the annotation here would silently break parsing.

// CabinClass uses UPPERCASE wire values (matches backend models.py:
// `class CabinClass(str, Enum): ECONOMY = "ECONOMY" ...`). The previous
// lowercase variants were producing 422s on POST /search/flights.
enum class CabinClass {
    @SerializedName("ECONOMY")
    ECONOMY,
    @SerializedName("PREMIUM_ECONOMY")
    PREMIUM_ECONOMY,
    @SerializedName("BUSINESS")
    BUSINESS,
    @SerializedName("FIRST")
    FIRST
}

enum class TripType {
    @SerializedName("roundtrip")
    ROUNDTRIP,
    @SerializedName("one_way")
    ONE_WAY,
    @SerializedName("multi_city")
    MULTI_CITY
}

enum class PriceLabel {
    @SerializedName("STEAL")
    STEAL,
    @SerializedName("GREAT_DEAL")
    GREAT_DEAL,
    @SerializedName("FAIR")
    FAIR,
    @SerializedName("EXPENSIVE")
    EXPENSIVE,
    @SerializedName("OVERPRICED")
    OVERPRICED,
    @SerializedName("UNKNOWN")
    UNKNOWN
}

enum class PriceTrend {
    @SerializedName("RISING")
    RISING,
    @SerializedName("FALLING")
    FALLING,
    @SerializedName("STABLE")
    STABLE,
    @SerializedName("VOLATILE")
    VOLATILE
}

enum class ActionType {
    @SerializedName("BUY_NOW")
    BUY_NOW,
    @SerializedName("WAIT")
    WAIT,
    @SerializedName("SET_ALERT")
    SET_ALERT,
    @SerializedName("MONITOR")
    MONITOR
}

// ===== Request/Response DTOs =====

/**
 * Mirrors backend `class SearchRequest(BaseModel)` in backend/models.py.
 * Field names below MUST match the Pydantic model exactly when serialized,
 * so each property carries an explicit @SerializedName. Don't add Kotlin
 * properties without a corresponding `@SerializedName` — Gson will encode
 * them in camelCase and the backend will 422 the request.
 */
data class SearchRequest(
    @SerializedName("origin")
    val origin: String,
    @SerializedName("destination")
    val destination: String,
    @SerializedName("departure_date")
    val departureDate: String,
    @SerializedName("return_date")
    val returnDate: String? = null,
    @SerializedName("adults")
    val adults: Int = 1,
    @SerializedName("children")
    val children: Int = 0,
    @SerializedName("infants")
    val infants: Int = 0,
    @SerializedName("cabin_class")
    val cabinClass: CabinClass = CabinClass.ECONOMY,
    @SerializedName("trip_type")
    val tripType: TripType = TripType.ROUNDTRIP,
    @SerializedName("max_results")
    val maxResults: Int = 20,
    // Backend field is `non_stop_only`. The previous wire name `direct_only`
    // was silently dropped on the server, which is why non-stop toggles in
    // the Android UI had no effect on results.
    @SerializedName("non_stop_only")
    val nonStopOnly: Boolean = false
)

data class IntentRequest(
    @SerializedName("query")
    val query: String
)

/**
 * Mirrors backend `class IntentResponse(BaseModel)`. The backend wire name
 * is `search_request` (not `parsed_request` — that was an inherited bug;
 * Gson would silently leave `parsedRequest` null).
 */
data class IntentResponse(
    @SerializedName("search_request")
    val searchRequest: SearchRequest,
    @SerializedName("confidence")
    val confidence: Float,
    @SerializedName("raw_query")
    val rawQuery: String,
    @SerializedName("interpretation")
    val interpretation: String
)

// ===== Response Models =====
//
// Stored properties + @SerializedName mirror the backend Pydantic schema
// exactly (see backend/models.py). Legacy Kotlin field names from earlier
// scaffold versions are preserved as computed properties (`val foo get() =
// ...`) so UI code (FlightCard, FlightDetailScreen, ParetoBar, etc.) keeps
// compiling without changes. Gson never sees the computed-property aliases.

/**
 * One leg of a flight (one takeoff → one landing). Mirrors backend Segment.
 * `departureAt` / `arrivalAt` arrive as ISO-8601 strings; the UI parses them
 * with `Instant.parse(...)` at render time, so we keep them as Strings here.
 */
data class Segment(
    @SerializedName("origin")
    val origin: String,
    @SerializedName("destination")
    val destination: String,
    @SerializedName("departure_at")
    val departureAt: String,
    @SerializedName("arrival_at")
    val arrivalAt: String,
    @SerializedName("carrier_code")
    val carrierCode: String,
    @SerializedName("flight_number")
    val flightNumber: String,
    @SerializedName("aircraft_code")
    val aircraftCode: String? = null,
    @SerializedName("duration_minutes")
    val durationMinutes: Int,
    @SerializedName("cabin")
    val cabin: CabinClass = CabinClass.ECONOMY
) {
    // ── Legacy aliases used throughout the UI ────────────────────────────
    val departureAirport: String get() = origin
    val arrivalAirport: String get() = destination
    val departureTime: String get() = departureAt
    val arrivalTime: String get() = arrivalAt
    val airline: String get() = carrierCode
    val aircraft: String get() = aircraftCode ?: ""
    // Backend Segment has no per-segment `stops` field (stops live on
    // Itinerary). Keeping this as 0 lets the existing UI loops compile.
    val stops: Int get() = 0
}

data class Itinerary(
    @SerializedName("segments")
    val segments: List<Segment>,
    @SerializedName("total_duration_minutes")
    val totalDurationMinutes: Int,
    @SerializedName("stops")
    val stops: Int = 0
) {
    // Legacy alias.
    val durationMinutes: Int get() = totalDurationMinutes
}

data class PriceBreakdown(
    @SerializedName("total_usd")
    val totalUsd: Double,
    @SerializedName("base_fare_usd")
    val baseFareUsd: Double,
    @SerializedName("taxes_usd")
    val taxesUsd: Double,
    @SerializedName("fees_usd")
    val feesUsd: Double = 0.0,
    @SerializedName("per_adult_usd")
    val perAdultUsd: Double? = null
) {
    // ── Legacy aliases ───────────────────────────────────────────────────
    val total: Double get() = totalUsd
    val baseFare: Double get() = baseFareUsd
    val basePrice: Double get() = baseFareUsd
    val taxes: Double get() = taxesUsd
    val fees: Double get() = feesUsd
}

data class BaggageInfo(
    @SerializedName("checked_bags_included")
    val checkedBagsIncluded: Int = 0,
    @SerializedName("carry_on_included")
    val carryOnIncluded: Boolean = true,
    @SerializedName("checked_bag_weight_kg")
    val checkedBagWeightKg: Int? = null
) {
    // ── Legacy aliases ───────────────────────────────────────────────────
    // FlightDetailScreen passes `checkedBags` to a Composable that expects
    // a String, so we surface it as a String here. Numeric callers should
    // read `checkedBagsIncluded` directly.
    val checkedBags: String get() = checkedBagsIncluded.toString()
    val carryOn: String get() = if (carryOnIncluded) "1 carry-on included" else "Not included"
}

data class FareConditions(
    @SerializedName("is_refundable")
    val isRefundable: Boolean = false,
    @SerializedName("change_fee_usd")
    val changeFeeUsd: Double? = null,
    @SerializedName("fare_class")
    val fareClass: String? = null
) {
    // ── Legacy aliases ───────────────────────────────────────────────────
    val refundable: Boolean get() = isRefundable
    // `changeable` is true if the backend supplies a change fee at all
    // (even $0); null means change-policy unknown. Matches iOS.
    val changeable: Boolean get() = changeFeeUsd != null
    // Old scaffold had this; backend doesn't surface it. Kept as a constant
    // so the UI doesn't error if it ever reads it.
    val seatSelectionIncluded: Boolean get() = false
}

data class PriceIntelligence(
    @SerializedName("price_label")
    val priceLabel: PriceLabel = PriceLabel.UNKNOWN,
    @SerializedName("price_percentile")
    val pricePercentile: Int? = null,
    @SerializedName("savings_vs_median_usd")
    val savingsVsMedianUsd: Double? = null,
    @SerializedName("savings_pct")
    val savingsPct: Double? = null,
    @SerializedName("trend")
    val trend: PriceTrend = PriceTrend.STABLE,
    @SerializedName("forecast_7d_usd")
    val forecast7dUsd: Double? = null,
    @SerializedName("forecast_14d_usd")
    val forecast14dUsd: Double? = null,
    @SerializedName("action")
    val action: ActionType = ActionType.MONITOR,
    @SerializedName("action_reason")
    val actionReason: String = "",
    @SerializedName("badge_text")
    val badgeText: String = "",
    @SerializedName("confidence")
    val confidence: Double = 0.0
) {
    // ── Legacy aliases ───────────────────────────────────────────────────
    // Existing UI code expects `percentile` as a non-null Int and uses it
    // for `progress = percentile / 100f` and "Cheaper than X% of fares"
    // copy. Backend ships `price_percentile: Int | null`, so default to 50
    // (median) when missing — keeps the progress bar centered.
    val percentile: Int get() = pricePercentile ?: 50
    val predicted7d: Double? get() = forecast7dUsd
    val predicted14d: Double? get() = forecast14dUsd
    val actionType: ActionType get() = action
}

data class FlightOffer(
    @SerializedName("offer_id")
    val offerId: String,
    @SerializedName("source")
    val source: String = "",
    @SerializedName("itineraries")
    val itineraries: List<Itinerary>,
    // Backend wire key is "price" (not "price_breakdown"). Keep the Kotlin
    // property name as `price` so existing `offer.price.total` call sites
    // work; legacy `priceBreakdown` is exposed as a computed alias.
    @SerializedName("price")
    val price: PriceBreakdown,
    // Wire key is "baggage"; UI reads it as `offer.baggageInfo`.
    @SerializedName("baggage")
    val baggage: BaggageInfo = BaggageInfo(),
    @SerializedName("fare_conditions")
    val fareConditions: FareConditions = FareConditions(),
    @SerializedName("seats_remaining")
    val seatsRemaining: Int? = null,
    @SerializedName("price_intelligence")
    val priceIntelligence: PriceIntelligence = PriceIntelligence(),
    @SerializedName("booking_url")
    val bookingUrl: String? = null,
    @SerializedName("last_ticketing_date")
    val lastTicketingDate: String? = null
) {
    // ── Identifiable / legacy aliases ────────────────────────────────────
    val id: String get() = offerId

    /** First slice — outbound leg of the trip. */
    val outboundItinerary: Itinerary
        get() = itineraries.firstOrNull()
            ?: Itinerary(segments = emptyList(), totalDurationMinutes = 0, stops = 0)

    /** Optional second slice — inbound (return) leg. Null for one-ways. */
    val returnItinerary: Itinerary? get() = itineraries.getOrNull(1)

    val baggageInfo: BaggageInfo get() = baggage

    // Backend now nests fare_class under fare_conditions.
    val fareClass: String get() = fareConditions.fareClass ?: ""

    /** Total stops summed across every itinerary in the offer. */
    val totalStops: Int get() = itineraries.sumOf { it.stops }
}

data class SearchResponse(
    @SerializedName("query_id")
    val queryId: String,
    @SerializedName("search_request")
    val searchRequest: SearchRequest,
    @SerializedName("offers")
    val offers: List<FlightOffer>,
    @SerializedName("total_found")
    val totalFound: Int = 0,
    @SerializedName("sources_queried")
    val sourcesQueried: List<String> = emptyList(),
    @SerializedName("returned_at")
    val returnedAt: String,
    @SerializedName("currency")
    val currency: String = "USD"
) {
    // ── Legacy aliases ───────────────────────────────────────────────────
    // Top-level route fields used to live on SearchResponse directly; the
    // backend now nests them under `search_request`. Expose flattened
    // accessors so ResultsScreen's `searchResponse.origin` etc. keep working.
    val searchId: String get() = queryId
    val origin: String get() = searchRequest.origin
    val destination: String get() = searchRequest.destination
    val departureDate: String get() = searchRequest.departureDate
    val returnDate: String? get() = searchRequest.returnDate
    val searchCompletedAt: String get() = returnedAt
}

File: ios/SkyAI/Models/Flight.swift
import Foundation

// MARK: - Enums

enum CabinClass: String, Codable, CaseIterable, Hashable {
    // Wire values are UPPERCASE to match backend Pydantic CabinClass
    // (see backend/models.py). Swift case names stay camelCase.
    case economy = "ECONOMY"
    case premiumEconomy = "PREMIUM_ECONOMY"
    case business = "BUSINESS"
    case first = "FIRST"

    var displayName: String {
        switch self {
        case .economy:
            return "Economy"
        case .premiumEconomy:
            return "Premium Economy"
        case .business:
            return "Business"
        case .first:
            return "First"
        }
    }
}

enum TripType: String, Codable, CaseIterable, Hashable {
    case roundtrip
    case oneWay = "one_way"
    case multiCity = "multi_city"

    var displayName: String {
        switch self {
        case .roundtrip:
            return "Round Trip"
        case .oneWay:
            return "One Way"
        case .multiCity:
            return "Multi-City"
        }
    }
}

enum PriceLabel: String, Codable {
    // Wire values UPPERCASE to match backend Pydantic PriceLabel.
    case steal = "STEAL"
    case greatDeal = "GREAT_DEAL"
    case fair = "FAIR"
    case expensive = "EXPENSIVE"
    case overpriced = "OVERPRICED"
    case unknown = "UNKNOWN"

    var displayName: String {
        switch self {
        case .steal:
            return "STEAL"
        case .greatDeal:
            return "GREAT DEAL"
        case .fair:
            return "FAIR"
        case .expensive:
            return "EXPENSIVE"
        case .overpriced:
            return "OVERPRICED"
        case .unknown:
            return "UNKNOWN"
        }
    }

    var badgeColor: String {
        switch self {
        case .steal:
            return "#FF6B35"
        case .greatDeal:
            return "#22C55E"
        case .fair:
            return "#9CA3AF"
        case .expensive:
            return "#F59E0B"
        case .overpriced:
            return "#EF4444"
        case .unknown:
            return "#D1D5DB"
        }
    }
}

enum PriceTrend: String, Codable {
    // Wire values UPPERCASE to match backend Pydantic PriceTrend.
    case rising = "RISING"
    case falling = "FALLING"
    case stable = "STABLE"
    case volatile = "VOLATILE"
}

enum ActionType: String, Codable {
    // Wire values UPPERCASE to match backend Pydantic ActionType.
    case buyNow = "BUY_NOW"
    case wait = "WAIT"
    case setAlert = "SET_ALERT"
    case monitor = "MONITOR"
}

// MARK: - Requests & Responses

struct IntentRequest: Codable {
    let query: String
}

struct IntentResponse: Codable {
    let searchRequest: SearchRequest
    let interpretation: String
    let confidence: Double

    enum CodingKeys: String, CodingKey {
        case searchRequest = "search_request"
        case interpretation
        case confidence
    }
}

struct SearchRequest: Codable, Equatable, Hashable {
    var originCode: String
    var destinationCode: String
    var departureDate: Date
    var returnDate: Date?
    var adults: Int
    var children: Int
    var infants: Int
    var cabinClass: CabinClass
    var tripType: TripType
    var directFlightsOnly: Bool
    var maxStops: Int?

    enum CodingKeys: String, CodingKey {
        // Wire format matches backend Pydantic SearchRequest (see backend/models.py).
        // Swift property names stay camelCase for readability; only JSON keys change.
        case originCode = "origin"
        case destinationCode = "destination"
        case departureDate = "departure_date"
        case returnDate = "return_date"
        case adults
        case children
        case infants
        case cabinClass = "cabin_class"
        case tripType = "trip_type"
        case directFlightsOnly = "non_stop_only"
        case maxStops = "max_stops"
    }

    init(
        originCode: String = "",
        destinationCode: String = "",
        departureDate: Date = Date().addingTimeInterval(86400),
        returnDate: Date? = nil,
        adults: Int = 1,
        children: Int = 0,
        infants: Int = 0,
        cabinClass: CabinClass = .economy,
        tripType: TripType = .roundtrip,
        directFlightsOnly: Bool = false,
        maxStops: Int? = nil
    ) {
        self.originCode = originCode
        self.destinationCode = destinationCode
        self.departureDate = departureDate
        self.returnDate = returnDate ?? Calendar.current.date(byAdding: .day, value: 30, to: departureDate)
        self.adults = adults
        self.children = children
        self.infants = infants
        self.cabinClass = cabinClass
        self.tripType = tripType
        self.directFlightsOnly = directFlightsOnly
        self.maxStops = maxStops
    }
}

// MARK: - Response Models
//
// Stored properties mirror the backend Pydantic schema exactly (see
// backend/models.py). Legacy Swift property names from earlier iOS
// versions are preserved as computed properties so UI code keeps
// compiling unchanged.

struct Segment: Codable {
    let origin: String
    let destination: String
    let departureAt: Date
    let arrivalAt: Date
    let carrierCode: String
    let flightNumber: String
    let aircraftCode: String?
    let durationMinutes: Int
    let cabin: CabinClass

    enum CodingKeys: String, CodingKey {
        case origin
        case destination
        case departureAt = "departure_at"
        case arrivalAt = "arrival_at"
        case carrierCode = "carrier_code"
        case flightNumber = "flight_number"
        case aircraftCode = "aircraft_code"
        case durationMinutes = "duration_minutes"
        case cabin
    }

    // Legacy API used by existing UI code.
    var departureAirport: String { origin }
    var arrivalAirport: String { destination }
    var departureTime: Date { departureAt }
    var arrivalTime: Date { arrivalAt }
    var airlineCode: String { carrierCode }
    var aircraft: String? { aircraftCode }
    // Backend Segment has no `stops` (stops live on Itinerary); UI still
    // reads this per-segment in a couple of places — keep it as 0.
    var stops: Int { 0 }

    var formattedDuration: String {
        let hours = durationMinutes / 60
        let minutes = durationMinutes % 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else {
            return "\(minutes)m"
        }
    }
}

struct Itinerary: Codable {
    let segments: [Segment]
    let totalDurationMinutes: Int
    let stops: Int

    enum CodingKeys: String, CodingKey {
        case segments
        case totalDurationMinutes = "total_duration_minutes"
        case stops
    }

    // Legacy name used by UI.
    var durationMinutes: Int { totalDurationMinutes }

    var formattedDuration: String {
        let hours = totalDurationMinutes / 60
        let minutes = totalDurationMinutes % 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else {
            return "\(minutes)m"
        }
    }

    var totalStops: Int { stops }
}

struct PriceBreakdown: Codable {
    let totalUsd: Double
    let baseFareUsd: Double
    let taxesUsd: Double
    let feesUsd: Double
    let perAdultUsd: Double?

    enum CodingKeys: String, CodingKey {
        case totalUsd = "total_usd"
        case baseFareUsd = "base_fare_usd"
        case taxesUsd = "taxes_usd"
        case feesUsd = "fees_usd"
        case perAdultUsd = "per_adult_usd"
    }

    // Legacy API.
    var basePrice: Double { baseFareUsd }
    var taxes: Double { taxesUsd }
    var fees: Double { feesUsd }
    var total: Double { totalUsd }
}

struct BaggageInfo: Codable {
    let checkedBagsIncluded: Int
    let carryOnIncluded: Bool
    let checkedBagWeightKg: Int?

    enum CodingKeys: String, CodingKey {
        case checkedBagsIncluded = "checked_bags_included"
        case carryOnIncluded = "carry_on_included"
        case checkedBagWeightKg = "checked_bag_weight_kg"
    }

    // Legacy API — UI expects these shapes.
    var carryon: String { carryOnIncluded ? "1 carry-on included" : "Not included" }
    var checkedBags: Int { checkedBagsIncluded }
    var checkedWeight: String? { checkedBagWeightKg.map { "\($0) kg" } }
}

struct FareConditions: Codable {
    let isRefundable: Bool
    let changeFeeUsd: Double?
    let fareClass: String?

    enum CodingKeys: String, CodingKey {
        case isRefundable = "is_refundable"
        case changeFeeUsd = "change_fee_usd"
        case fareClass = "fare_class"
    }

    // Legacy API. `changeable` is true if backend supplies a change fee
    // (even $0), false if the fee is unspecified/nil.
    var refundable: Bool { isRefundable }
    var changeable: Bool { changeFeeUsd != nil }
    var minStayDays: Int? { nil }
}

struct PriceIntelligence: Codable {
    let priceLabel: PriceLabel
    let pricePercentile: Int?
    let savingsVsMedianUsd: Double?
    let savingsPct: Double?
    let trend: PriceTrend
    let forecast7dUsd: Double?
    let forecast14dUsd: Double?
    let action: ActionType
    let actionReason: String
    let badgeText: String
    let confidence: Double

    enum CodingKeys: String, CodingKey {
        case priceLabel = "price_label"
        case pricePercentile = "price_percentile"
        case savingsVsMedianUsd = "savings_vs_median_usd"
        case savingsPct = "savings_pct"
        case trend
        case forecast7dUsd = "forecast_7d_usd"
        case forecast14dUsd = "forecast_14d_usd"
        case action
        case actionReason = "action_reason"
        case badgeText = "badge_text"
        case confidence
    }

    // Legacy API — mapped to new fields so UI keeps working.
    var percentileRank: Double { Double(pricePercentile ?? 50) / 100.0 }
    var savingsAmount: Double? { savingsVsMedianUsd }
    var savingsPercent: Double? { savingsPct }
    var recommendedAction: ActionType { action }
    var explanation: String { actionReason }
    // Backend no longer sends these; kept nil for UI compatibility.
    var historicalAverage: Double? { nil }
    var historicalMin: Double? { nil }
    var historicalMax: Double? { nil }
}

struct FlightOffer: Codable, Identifiable {
    let offerId: String
    let source: String
    let itineraries: [Itinerary]
    let priceBreakdown: PriceBreakdown     // decoded from backend "price"
    let baggageInfo: BaggageInfo           // decoded from backend "baggage"
    let fareConditions: FareConditions
    let seatsRemaining: Int?
    let priceIntelligence: PriceIntelligence
    let bookingUrl: String?
    let lastTicketingDate: Date?

    enum CodingKeys: String, CodingKey {
        case offerId = "offer_id"
        case source
        case itineraries
        case priceBreakdown = "price"
        case baggageInfo = "baggage"
        case fareConditions = "fare_conditions"
        case seatsRemaining = "seats_remaining"
        case priceIntelligence = "price_intelligence"
        case bookingUrl = "booking_url"
        case lastTicketingDate = "last_ticketing_date"
    }

    // Identifiable conformance.
    var id: String { offerId }

    // Legacy API.
    var outbound: Itinerary {
        itineraries.first ?? Itinerary(segments: [], totalDurationMinutes: 0, stops: 0)
    }
    var inbound: Itinerary? {
        itineraries.count > 1 ? itineraries[1] : nil
    }
    var availableSeats: Int { seatsRemaining ?? 0 }

    // UI helpers (signature unchanged).
    var totalDuration: String {
        let totalMinutes = outbound.durationMinutes + (inbound?.durationMinutes ?? 0)
        let hours = totalMinutes / 60
        let minutes = totalMinutes % 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else {
            return "\(minutes)m"
        }
    }

    var totalStops: Int {
        itineraries.reduce(0) { $0 + $1.stops }
    }

    var priceLabel: String {
        priceIntelligence.priceLabel.displayName
    }

    var price: Double {
        priceBreakdown.totalUsd
    }
}

struct SearchResponse: Codable {
    let queryId: String
    let searchRequest: SearchRequest
    let offers: [FlightOffer]
    let totalFound: Int
    let sourcesQueried: [String]
    let returnedAt: Date
    let currency: String

    enum CodingKeys: String, CodingKey {
        case queryId = "query_id"
        case searchRequest = "search_request"
        case offers
        case totalFound = "total_found"
        case sourcesQueried = "sources_queried"
        case returnedAt = "returned_at"
        case currency
    }

    // Legacy name.
    var searchTimestamp: Date { returnedAt }
}
